packages/contracts/contracts/L2/ValidatorRewardVault.sol

```diff
@@ -11,10 +14,153 @@ import { Semver } from "../universal/Semver.sol";
+    }
+
+    /**
+     * @notice Rewards the validator for submitting the output.
+     *         ValidatorPool contract on L1 calls this function over the portal when output is finalized.
+     *         As the submission of outputs is delayed, the amount of rewards decreases.
+     *
+     * @param _validator     Address of the validator.
+     * @param _l2BlockNumber The L2 block number of the output root.
+     * @param _penaltyNum    The penalty period which reduces the reward amount.
+     * @param _penaltyDenom  The maximum value of the penalty period.
+     */
+    function reward(
+        address _validator,
+        uint256 _l2BlockNumber,
+        uint256 _penaltyNum,
+        uint256 _penaltyDenom
+    ) external {
+        require(
+            AddressAliasHelper.undoL1ToL2Alias(msg.sender) == VALIDATOR_POOL,
+            "ValidatorRewardVault: function can only be called from the ValidatorPool"
+        );
+
+        require(_validator != address(0), "ValidatorRewardVault: validator address cannot be 0");
+
+        require(
+            !isPaid[_l2BlockNumber],
+            "ValidatorRewardVault: the reward has already been paid for the L2 block number"
+        );
+
+        uint256 amount = _determineRewardAmount(_penaltyNum, _penaltyDenom);
+        totalReserved += amount;
+        rewards[_validator] += amount;
+        isPaid[_l2BlockNumber] = true;
+
+        emit Rewarded(_validator, _l2BlockNumber, amount);
+    }
+
+    /**
+     * @notice Withdraws all of the sender's balance.
+     *         Reverts if the balance is less than the minimum withdrawal amount.
+     */
+    function withdraw() external override {
+        uint256 balance = rewards[msg.sender];
+
+        require(
+            balance >= MIN_WITHDRAWAL_AMOUNT,
+            "ValidatorRewardVault: withdrawal amount must be greater than minimum withdrawal amount"
+        );
+
+        rewards[msg.sender] = 0;
+
+        unchecked {
+            totalReserved -= balance;
+            totalProcessed += balance;
+        }
+
+        emit Withdrawal(balance, msg.sender, msg.sender);
+
+        L2StandardBridge(payable(Predeploys.L2_STANDARD_BRIDGE)).bridgeETHTo{ value: balance }(
+            msg.sender,
+            WITHDRAWAL_MIN_GAS,
+            bytes("")
+        );
+    }
+
+    /**
+     * @notice Determines the reward amount penalized.
+     *         Note that if the penalty is 0, no penalty is applied.
+     *
+     * @param _penaltyNum   The penalty period which reduces the reward amount.
+     * @param _penaltyDenom The maximum value of the penalty period.
+     *
+     * @return Amount of the reward.
+     */
+    function _determineRewardAmount(uint256 _penaltyNum, uint256 _penaltyDenom)
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 fullReward = (address(this).balance - totalReserved) / REWARD_DIVIDER;
+
+        if (_penaltyDenom == 0) {
+            return fullReward;
+        }
```

Comment:
Is it for gas optimization? Because the result will be same with or without this line.

Reply:
If penaltyDenom is 0, a division by 0 error occurs.

Reply:
Ah okay, then is this the case for `revert()`? If you think not, please move up to the start of the function body, as it's a function argument validation part.